nifi-commons/nifi-record/src/main/java/org/apache/nifi/serialization/record/util/DataTypeUtils.java

```diff
@@ -469,21 +473,51 @@ public static RecordSchema inferSchema(final Map<String, Object> values, final S
             final RecordField recordField = new RecordField(key, inferredDataType, true);
             inferredFieldTypes.add(recordField);
 
-            final Object coercedValue = convertType(rawValue, inferredDataType, fieldName, charset);
-            coercedValues.put(key, coercedValue);
+            convertType(rawValue, inferredDataType, fieldName, charset);
         }
 
-        final RecordSchema inferredSchema = new SimpleRecordSchema(inferredFieldTypes);
-        return inferredSchema;
+        return new SimpleRecordSchema(inferredFieldTypes);
     }
 
     public static Record toRecord(final Object value, final String fieldName, final Charset charset) {
+        return toRecord(value, fieldName, charset, false);
+    }
+
+    private static Object covertObjectToRecord(final Object rawValue, final String key, final Charset charset) {
+        final Object coercedValue;
+        if (rawValue instanceof Map<?, ?>) {
+            coercedValue = toRecord(rawValue, key, charset, true);
+        } else if (rawValue instanceof Object[]) {
+            final Object[] objArray = (Object[]) rawValue;
+            coercedValue = Arrays.stream(objArray).noneMatch(o -> o instanceof Map<?, ?>)
+                    ? objArray
+                    : Arrays.stream(objArray).map(o -> toRecord(o, key, charset, true)).toArray();
+        } else if (rawValue instanceof Collection<?>) {
+            final Collection<?> objCollection = (Collection<?>) rawValue;
+            // Records have ARRAY DataTypes, so convert any Collections
+            coercedValue = objCollection.stream().noneMatch(o -> o instanceof Map<?, ?>)
+                    ? objCollection.toArray()
+                    : objCollection.stream().map(o -> toRecord(o, key, charset, true)).toArray();
```

Comment:
We should avoid creating a `Stream` in this part of the codebase, as it is very sensitive to performance.

Reply:
That's fair, I've reverted this and found a coupleof other Streams within `DataTypeUtils` that should probably be changed too, will include in this PR